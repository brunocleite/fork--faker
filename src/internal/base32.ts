export const reducedBase32 = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'; // Crockford's Base32 - Excludes I, L, O, and U which may be confused with numbers

/**
 * Encodes a Date into 10 characters base32 string.
 * @param date the Date to encode
 */
export const encodeDate = (date: Date): string => {
  let now = date.getTime();

  let mod;
  let len = 10;
  let str = '';
  for (; len > 0; len--) {
    mod = now % reducedBase32.length;
    str = reducedBase32.charAt(mod) + str;
    now = (now - mod) / reducedBase32.length;
  }

  return str;
};
