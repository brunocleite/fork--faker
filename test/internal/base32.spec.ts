import { describe, expect, it } from 'vitest';
import { encodeDate, reducedBase32 } from '../../src/internal/base32';

describe('encodeDate()', () => {
  it('encodes current date correctly', () => {
    const date = new Date('2023-04-01T00:00:00Z');
    const encoded = encodeDate(date);
    expect(encoded).toHaveLength(10);
    expect(encoded).toMatchSnapshot();
    for (const char of encoded)
      expect(reducedBase32.includes(char)).toBeTruthy();
  });

  it('encodes epoch start date correctly', () => {
    const date = new Date('1970-01-01T00:00:00Z');
    const encoded = encodeDate(date);
    expect(encoded).toBe('0000000000');
  });

  it('returns different encodings for dates one millisecond apart', () => {
    const date1 = new Date('2023-04-01T00:00:00.000Z');
    const date2 = new Date('2023-04-01T00:00:00.001Z');
    const encoded1 = encodeDate(date1);
    const encoded2 = encodeDate(date2);
    expect(encoded1).not.toBe(encoded2);
  });

  it('encodes same date consistently', () => {
    const date = new Date('2023-04-01T00:00:00Z');
    const encoded1 = encodeDate(date);
    const encoded2 = encodeDate(date);
    expect(encoded1).toBe(encoded2);
  });
});
