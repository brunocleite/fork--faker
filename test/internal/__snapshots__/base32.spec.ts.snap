// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encodeDate() > encodes current date correctly 1`] = `"01GWX1T800"`;
